pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -204,32 +269,251 @@ func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconci
+					return err
+				}
+		}
+		return nil
+}
+
+func (r *ReconcileNodeMaintenance) obtainLease(node *corev1.Node) (error, bool) {
+	if !r.isLeaseSupported {
+		return nil, false
+	}
+
+	log.Info("Lease object supported, obtaining lease")
+	lease, needUpdate, err := r.createOrGetExistingLease(node, LeaseDurationInSeconds)
 
-	// Uncordon the node
-	err = AddOrRemoveTaint(r.drainer.Client, node, false)
 	if err != nil {
-		return err
+		log.Errorf("failed to create or get existing lease error=%v", err)
+		return err, false
 	}
 
-	if err = drain.RunCordonOrUncordon(r.drainer, node, false); err != nil {
-		return err
+	if needUpdate {
+
+		log.Info("update lease - lease duration change requested")
+
+		if _, err, updateOwnedLeaseFailed := updateLease(r.client, node, lease, time.Now(), LeaseDurationInSeconds); err != nil {
+				return err, updateOwnedLeaseFailed
+		}
 	}
 
-	return nil
+	return nil, false
+}
+
+
+func makeInt32(val int32) (*int32) {
+	tmpVal := val
+	return &tmpVal
+}
+
+func makeTimeNow(time time.Time) (*metav1.MicroTime) {
+
+	timeNow := metav1.MicroTime{Time: time}
+	return &timeNow
+}
+
+func makeExpectedOwnerOfLease(node *corev1.Node) (*metav1.OwnerReference) {
+	return &metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       node.ObjectMeta.Name,
+		UID:        node.ObjectMeta.UID,
+	}
+}
+
+
+func (r *ReconcileNodeMaintenance) createOrGetExistingLease(node *corev1.Node, durationInSeconds int32) (*coordv1beta1.Lease, bool, error) {
+
+	holderIdentity := LeaseHolderIdentity
+	owner := makeExpectedOwnerOfLease(node)
+	microTimeNow :=  makeTimeNow(time.Now())
+
+	lease := &coordv1beta1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            node.ObjectMeta.Name,
+			Namespace:       LeaseNamespace,
+			OwnerReferences: []metav1.OwnerReference{*owner},
+		},
+		Spec: coordv1beta1.LeaseSpec{
+			HolderIdentity:       &holderIdentity,
+			LeaseDurationSeconds: makeInt32(durationInSeconds),
+			AcquireTime:          microTimeNow,
+			RenewTime:            microTimeNow,
+			LeaseTransitions:     makeInt32(0),
+		},
+	}
+
+	if err := r.client.Create(context.TODO(), lease); err != nil {
+		if errors.IsAlreadyExists(err) {
+
+				nodeName := node.ObjectMeta.Name
+				nName := apitypes.NamespacedName{Namespace: LeaseNamespace, Name: nodeName}
+
+				if err := r.client.Get(context.TODO(), nName, lease); err != nil {
+					return  nil, false, err
+				}
+				return lease, true, nil
+		}
+		return nil, false, err
+	}
+	return lease, false,  nil
+}
+
+func needUpdateOwnedLease(lease *coordv1beta1.Lease, currentTime time.Time) (bool,bool) {
+
+	if lease.Spec.RenewTime == nil || lease.Spec.LeaseDurationSeconds == nil {
+		log.Info("empty renew time or duration in sec")
+		return true, true
+	}
+	dueTime := lease.Spec.RenewTime.Time
+	duration := *lease.Spec.LeaseDurationSeconds
+	dueTime = dueTime.Add(time.Duration(time.Duration(duration) * time.Second))
```

Comment:
redundant `time.Duration(...)`

Reply:
doesn't compile, have to cast from int32

Reply:
is that better? can't get around the cast because the struct field is *int32
```
	duration := time.Duration(*lease.Spec.LeaseDurationSeconds)
	dueTime = dueTime.Add(duration * time.Second)
```

Reply:
yes :+1: